### PR TITLE
CCv0: Attestation Agent Integration

### PR DESCRIPTION
**Pull request by fitzthum**
Use the Attestation Agent as a keyprovider for Skopeo/ocicrypt when we pull the container from inside the guest. This requires starting the Attestation Agent, creating an ocicrypt config file, and amending the arguments of the skopeo copy command.

### CODE REVIEW

**Comment by fitzthum** [comment]
What exactly is the aa_kbc_kbs parameter? I thought the `kbc_name` and `kbs_uri` were both part of the payload sent by ocicrypt.

**Comment by jialez0** [comment]
@fitzthum This KBC_KBS information is what we call KBC name and KBS URI. Their current format is like `xxx_kbc::<KBS_URI>`. Do you remember? The result of our previous discussion is that these two information may be placed in the configuration file of kata agent, passed by kata agent to ocicrypt-rs, and then passed to AA through keyprovider protocol. in the current ccv0 scheme, this parameter is actually passed directly through the provider parameter in the `skopeo copy` command.

**Comment by fitzthum** [comment]
@jialez0 Are you sure about that? As I recall we decided to move the `kbc_name` and `kbs_uri` from the annotation, but @voidC1326 and @liangzhou121 wanted it to be part of the keyprovider_protocol rather than have it be a startup parameter of the AA. This is in the discussion here https://github.com/containers/attestation-agent/pull/5

Furthermore, if you look at the source, it seems like the `kbc_name` and `kbs_uri` come from the `un_wrap_key` function in `src/grpc/mod.rs` where they are derived from the gRPC request. Am I missing something?

**Comment by jialez0** [comment]
@fitzthum Yeah, that's exactly what I mean. `kbc_name` and `kbs_uri` here are extra parameters from the provider of skopeo, then ocicrypt will add them to keyprovider request. 
PS: My previous GitHub ID was voidC1326, but now it's jialez0 :)

**Comment by fitzthum** [comment]
@jialez0 Good question. We could put `kbc_name` and `kbs_uri` in the agent config. That would mean that the measurement would change each time either of those parameters changes. You would basically need to provide a new initrd each time you move the KBS and two tenants couldn't share the same config. I think that's why we were talking about putting it into the pod annotation. I don't think either of those parameters needs to be measured (a malicious KBS shouldn't be able to do much) and having them be set by the pod annotation might make orchestration easier.

I'm not totally sure what that would entail from an engineering perspective, but I will look into it.

**Comment by fitzthum** [comment]
Btw I will rebase this with the various changes and post a cleaner version in the nearish future.

**Comment by jialez0** [comment]
@fitzthum Okey, my little worry is the engineering feasibility of putting it in the pod annotation. As far as I know, the agent currently reads the pod annotation when the `StartContainer` is called. If we put KBC name and KBS URI in the pod annotation, we may need to advance this reading task to `PullImage`?

**Comment by fitzthum** [comment]
Yeah setting up the parameters via the annotation is definitely more complicated. I guess we can skip that for CCv0 and just use the config file, but let's keep it in mind for the future.

File: src/agent/src/rpc.rs [hunk 1]
```diff
@@ -1737,6 +1774,137 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+    use_aa: bool,
+    agent_started: bool,
+) -> Result<()> {
+    let source_image = format!("{}{}", "docker://", image);
+
+    let manifest_path = format!("/tmp/{}/image_manifest", cid);
+    let target_path_manifest = format!("dir://{}", manifest_path);
+
+    // Define the target transport and path for the OCI image, without signature
+    let oci_path = format!("/tmp/{}/image_oci:latest", cid);
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:1790]: - Can you make `image_oci:latest` a `const` as it's hard-coded multiple times?
- Would there ever be a need to *not* use latest? We often see problems with the latest versions so it might be prudent to allow the specification of an alternative image somehow.

File: src/agent/src/rpc.rs [hunk 2]
```diff
@@ -1737,6 +1774,137 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+    let target_path_oci = format!("oci://{}", oci_path);
+
+    fs::create_dir_all(&manifest_path)?;
+    fs::create_dir_all(&oci_path)?;
+
+    info!(sl!(), "Attempting to pull image {}...", &source_image);
+
+    let mut pull_command = Command::new(SKOPEO_PATH);
+    pull_command
+        .arg("--insecure-policy")
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:1800]: Please add a comment explaining why this option is needed.

File: src/agent/src/rpc.rs [hunk 3]
```diff
@@ -1737,6 +1774,137 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+        }
+
+        // not 100% sure we actually need this. the arg gets ignored anyway
+        pull_command
+            .arg("--decryption-key")
+            .arg("provider:attestation-agent:offline_fs_kbc::null")
+            .env("OCICRYPT_KEYPROVIDER_CONFIG", OCICRYPT_CONFIG_PATH);
+
+        //TODO: check if image is really encrypted since skopeo doesn't
+        //      seem to care
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:1821]: Needs a GitHub issue URL for tracking this remaining item.

File: src/agent/src/rpc.rs [hunk 4]
```diff
@@ -1737,6 +1774,137 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+
+    if use_aa {
+        if !agent_started {
+            init_attestation_agent();
+        }
+
+        // not 100% sure we actually need this. the arg gets ignored anyway
+        pull_command
+            .arg("--decryption-key")
+            .arg("provider:attestation-agent:offline_fs_kbc::null")
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:1817]: ? Maybe a comment would be helpful here too?

File: src/agent/src/rpc.rs [hunk 5]
```diff
@@ -1737,6 +1774,137 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+    let mut config_file = File::create(config_path).unwrap();
+    config_file
+        .write_all(ocicrypt_config.to_string().as_bytes())
+        .unwrap();
+
+    // start AA
+    // should we use tokio for this?
+    Command::new(AA_PATH)
+        .arg("--grpc_sock")
+        .arg(format!("127.0.0.1:{}", AA_PORT))
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:1901]: Can you create a `const for `127.0.0.1` which is also hard-coded multiple times?

File: src/agent/src/rpc.rs [hunk 6]
```diff
@@ -1737,6 +1774,137 @@ fn load_kernel_module(module: &protocols::agent::KernelModule) -> Result<()> {
...
+        .spawn()
+        .unwrap();
+
+    //TODO: we should probably clean up the AA on shutdown?
```
Comment (jodh-intel) [inline src/agent/src/rpc.rs:1905]: - How do we determine the answer to this?
- We need to ensure the shutdown is not blocked indefinitely for the agent tracing scenario.

**Review by jodh-intel — COMMENTED** [review]
Thanks @fitzthum.

Some comments (some the same as the various other implementations floating around ;)

**Comment by jodh-intel** [comment]
Any update on this @fitzthum? Please can you rebase to resolve the conflicts too.

File: src/agent/src/rpc.rs (hunk 1 above)
Comment (fitzthum) [reply src/agent/src/rpc.rs:1790]: Sure, this is not part of my patch tho. I guess the rebase has kind of confounded this PR.

File: src/agent/src/rpc.rs (hunk 2 above)
Comment (fitzthum) [reply src/agent/src/rpc.rs:1800]: Again, not part of my patch.

File: src/agent/src/rpc.rs (hunk 3 above)
Comment (fitzthum) [reply src/agent/src/rpc.rs:1821]: I've been thinking about this a little bit, and it probably isn't worth it to add anything to Kata here. Instead, I will open an issue on the new `image-rs` project to make sure that copying an image will fail if you provide a keyprovider but the image is not encrypted.

File: src/agent/src/rpc.rs (hunk 1 above)
Comment (jodh-intel) [reply src/agent/src/rpc.rs:1790]: Ack. As mentioned on another #2658, imho we should try to get these CC PRs landed into the `CCv0` branch as soon as we can, close outstanding stale PRs and potentially break the PRs into smaller units to speed up the process. I understand how we've got to the current situation, but it's starting to get extremely confusing, atleast to me.

**Comment by fitzthum** [comment]
Yeah I rebase and try to get this PR back to just the commits that it started as.

**Comment by jodh-intel** [comment]
Ok, thanks @fitzthum.

File: src/agent/src/rpc.rs (hunk 6 above)
Comment (Jakob-Naucke) [reply src/agent/src/rpc.rs]: @fitzthum do we have any cleanup functionality/tasks whatsoever?

File: src/agent/src/image_rpc.rs [hunk 7]
```diff
@@ -121,10 +140,44 @@ impl ImageService {
...
+            .unwrap();
+
+        // The Attestation Agent will run for the duration of the guest.
+        Command::new(AA_PATH)
+            .arg("--grpc_sock")
+            .arg(AA_PORT)
+            .spawn()
+            .unwrap();
+
+        //TODO: Do we need to clean up the AA on shutdown?
```
Comment (Jakob-Naucke) [inline src/agent/src/image_rpc.rs:170]: @fitzthum do you want to keep this a TODO for this PR?

**Comment by wllenyj** [comment]
There is a config `image.provision` about Container image decryption keys provisioning in the configuration-qemu.toml.in. 
It is added by this patch https://github.com/kata-containers/kata-containers/commit/539afba03daa341474a6223a852bed09eb835bc4. 
Does it have anything to do with skopeo's decryption-key params ?

File: src/agent/src/image_rpc.rs [hunk 8]
```diff
@@ -121,10 +140,44 @@ impl ImageService {
...
+    }
+
     async fn pull_image(&self, req: &image::PullImageRequest) -> Result<String> {
         let image = req.get_image();
         let mut cid = req.get_container_id();
 
+        let agent_config = AGENT_CONFIG.read().await;
+        let aa_kbc_params = &agent_config.aa_kbc_params;
+        let aa_started = agent_config.confidential_setup_complete;
```
Comment (wllenyj) [inline src/agent/src/image_rpc.rs:179]: `let aa_started = !agent_config.confidential_setup_complete;` ?
Or change the naming?

File: src/agent/src/image_rpc.rs [hunk 9]
```diff
@@ -136,9 +189,15 @@ impl ImageService {
             verify_cid(cid)?;
         }
 
+        if aa_kbc_params != "" && aa_started {
```
Comment (wllenyj) [inline src/agent/src/image_rpc.rs:192]: `! aa_started` ?  the same as above
and `aa_kbc_params.is_empty()`. Otherwise static-check will fail.

**Comment by Jakob-Naucke** [comment]
> There is a config `image.provision` about Container image decryption keys provisioning in the configuration-qemu.toml.in. It is added by this patch [539afba](https://github.com/kata-containers/kata-containers/commit/539afba03daa341474a6223a852bed09eb835bc4). Does it have anything to do with skopeo's decryption-key params ?

@wllenyj yeah, just recently I was reminded about these options again. We discussed them in #1332 when much about how things would work wasn't really clear.

- `data:///local/key/file`: I think this was added for some vague idea of what would later become the [offline file system key broker client](https://github.com/confidential-containers/attestation-agent/tree/main/src/kbc_modules/offline_fs_kbc) (read from a file, which is secure when the initrd is encrypted, e.g. Secure Execution, PEF IIUC). However, the offline FS KBC is an Attestation Agent module, so this isn't very useful.
- `file:///path/to/bin/fetcher/in/guest`: This _seems_ what the Attestation Agent is now _at first glance_, but I think it was more intended for firmware calls, which is somewhat related to SEV{,-ES}'s pre-attestation. This is currently under PR as an Attestation Agent module, https://github.com/confidential-containers/attestation-agent/pull/19.
- `https://my-key-broker.foo/tenant/<tenant-id>`: This was intended for remote key provision, e.g. through runtime attestation with EAA ([KBC module](https://github.com/confidential-containers/attestation-agent/tree/main/src/kbc_modules/eaa_kbc)), TDX, SEV-SNP. (One might also think about providing keys remotely through static TLS certificates in encrypted images, which would be a bit of a hybrid with the first way).

In short, these configurations are superseded by the Attestation Agent, and in some cases, it might actually be important to have these things measured, i.e. they shouldn't be host-configurable in the first place like this configuration skeleton suggests.
They are generally superseded by the `aa_kbc_params` variable suggested here (which is guest, not host configuration) and the architecture of the respective AA KBC modules.

@sameo am I overlooking something here?

File: src/agent/src/image_rpc.rs (hunk 7 above)
Comment (fitzthum) [reply src/agent/src/image_rpc.rs:170]: I can remove it given that it is not really significant for CCv0. I think there is some discussion to be had about the best way to close persistent connections that the AA has started. If we just use the AA for getting container secrets all at once and at the beginning of execution, then we can probably just have the server timeout and not worry about cleaning anything up for the AA.

**Comment by Jakob-Naucke** [comment]
> Is it possible for AA to be started by Systemd, same as `agent` ?

@wllenyj Could be. However, I think we should not cease supporting what is suggested in this PR as long as #2386 is not resolved, as some architectures depend on initrd to run in a TEE.

File: src/agent/src/image_rpc.rs (hunk 8 above)
Comment (fitzthum) [reply src/agent/src/image_rpc.rs:178]: If the setup is complete the agent has been started, so I think this name is fine. I changed the `!` below.

**Comment by fitzthum** [comment]
> > Is it possible for AA to be started by Systemd, same as `agent` ?
> 
> @wllenyj Could be. However, I think we should not cease supporting what is suggested in this PR as long as #2386 is not resolved, as some architectures depend on initrd to run in a TEE.

We haven't talked much about guest initrd support for CC, but probably we want to support guests that run the agent as init and don't have systemd.

File: src/agent/src/image_rpc.rs (hunk 8 above)
Comment (wllenyj) [reply src/agent/src/image_rpc.rs:178]: Oh, yes,  my mistake,  naming is ok.

File: src/agent/src/config.rs [hunk 10]
```diff
@@ -180,6 +185,7 @@ impl FromStr for AgentConfig {
         config_override!(agent_config_builder, agent_config, server_addr);
         config_override!(agent_config_builder, agent_config, unified_cgroup_hierarchy);
         config_override!(agent_config_builder, agent_config, tracing);
+        config_override!(agent_config_builder, agent_config, aa_kbc_params);
```
Comment (wllenyj) [inline src/agent/src/config.rs:188]: Do we need support to configure this parameter separately? It seems that it can only be configured via config_file now.

Comment (Jakob-Naucke) [reply src/agent/src/config.rs:188]: @wllenyj that is what this macro does, it sets the variable to what was set in the config file. See #2517 for the introduction of this.

**Comment by Jakob-Naucke** [comment]
@liangzhou121 did you use this PR for last Thursday's demo? When I try it, the log messages stop, no pull is happening, and later it looks as if the attestation agent is being started with another instance of it already running.

**Comment by fitzthum** [comment]
I think they are using this PR, but only to get the parameters from the config file. I think the demo was using the rust agent patch and skipping over any skopeo-specific encryption stuff.

**Comment by sameo** [comment]
@fitzthum Could we rebase that PR? Or do you want to wait for PR #3035 to be merged first?

File: src/agent/src/image_rpc.rs [hunk 11]
```diff
@@ -136,9 +188,15 @@ impl ImageService {
             verify_cid(cid)?;
         }
 
+        if aa_kbc_params != "" && !aa_started {
+            Self::init_attestation_agent();
+            let mut agent_config = AGENT_CONFIG.write().await;
```
Comment (Jakob-Naucke) [inline src/agent/src/image_rpc.rs:193]: @sameo we see this lock failing to be acquired with shim pull, would you have an idea why that happens?

Comment (fitzthum) [reply src/agent/src/image_rpc.rs]: New version sidesteps the agent config by adding an AtomicBool as a field in the image service.

**Comment by Jakob-Naucke** [comment]
I can get this to work with Secure Execution, https://github.com/confidential-containers/documentation/pull/7 works. Execs are blocked etc.

**Comment by liangzhou121** [comment]
> 

@fitzthum @Jakob-Naucke Yes, we only applied this PR partially (only agent config file load changes) in order to pass EAA KBC parameters in our demo. So we didn't encounter the strange error you mentioned.

**Comment by fitzthum** [comment]
> I can get this to work with Secure Execution, [confidential-containers/documentation#7](https://github.com/confidential-containers/documentation/pull/7) works. Execs are blocked etc.

Works with SEV as well.

**Comment by Jakob-Naucke** [comment]
@sameo anything you think needs to be addressed before Thanksgiving?

**Review by sameo — APPROVED** [review]

**Comment by sameo** [comment]
> @sameo anything you think needs to be addressed before Thanksgiving?

It LGTM. The only failing check is the snap one and that's because the snap servers seem to be having a bad day. I'm going to merge that one regardless.
